### PR TITLE
Convert data-broker to TS

### PR DESCRIPTION
**Pull request by kael89**
### Issue #:
[No issue]

This PR adds types for `data-broker`! I tried to make as few actual code changes as possible in this one.

Sorry, this is a big PR! But I thought that it makes sense for this kind of change. Plus, it will be easier for me to manage now that I am a member of the Open Source Community😉 

Note that the build will fail for various reasons:
1. There are some TS errors, I will fix those in the next PR. The tests will need significant changes in some places. Errors in production code are less (<10). I can make some changes that _should_ be safe from regression. or use directives like `ts-ignore`/`ts-expect-error` to suppress them.
2. There is a circular dependency between `data-broker` and `indicators`. This was already in place but it now makes the `yarn build` command in `data-broker` fail in my local, if I also build `indicators`. I don't know if this will happen in CI/CD, or if it will still be there after all TS errors are addressed, but let's keep an eye on it.

Point 1 above is based on the assumption that CI/CD builds fail now for TS errors, but I'm not actually sure if this is true

### CODE REVIEW

**Comment by edmofro** [comment]
Hooray for the Open Source Community!!! This is a genuinely super helpful improvement @kael89, thank you so so much for it 🙏 🙏 🙏 
Will post it in the reviews channel in slack so it gets picked up in the normal way (and you get some well deserved public credit 😉 )

**Comment by IgorNadj** [comment]
Hey Kostas! Good to hear from ya again :)

>There is a circular dependency between data-broker and indicators

Yeah there's a few in our dependencies! I have a PR to clean these up. Currently dependency loops don't fail our builds. But if any one individual package build fails, it will fail the build.

Here is how I untangled this loop: https://github.com/beyondessential/tupaia/pull/3931/commits/041391bbbefca5fc843c36e13df5a96ea3b78140

File: packages/data-broker/src/types/results.ts
Comment (IgorNadj) [reply packages/data-broker/src/types/results.ts:6]: It's ok, we are creating a @tupaia/types package soon that will DRY this.

File: packages/data-broker/src/services/dhis/pullers/DeprecatedEventsPuller.ts
Comment (IgorNadj) [inline packages/data-broker/src/services/dhis/pullers/DeprecatedEventsPuller.ts:9]: I've seen this Partial pattern used quite a bit in this PR. What's the advantage of Partial over the old school optional key syntax `{ a?: number }`. I like the old school way because it lets us do a mix of required and optional keys.

File: packages/data-broker/src/services/tupaia/TupaiaService.ts
Comment (IgorNadj) [inline packages/data-broker/src/services/tupaia/TupaiaService.ts:1]: Looks like git is pretty smart and can track most of these renames, but not all. It would be nice to keep the history for git blame, and to do that we would need to split up this commit, one to rename the files with no changes, and the second to add the types. 

I think it's still worth doing and not too hard to split that up. If you do a new commit with just the renames and then cherry-pick this commit on top that should do the trick.

File: packages/data-broker/src/services/weather/types.ts
Comment (IgorNadj) [inline packages/data-broker/src/services/weather/types.ts:14]: Same comment as before about unioned config type.

File: packages/data-broker/src/types/models.ts
Comment (IgorNadj) [reply packages/data-broker/src/types/models.ts:33]: Nice one. Looks like we probably want to move this to @tupaia/database or @tupaia/types at some point.

File: packages/data-broker/src/types/models.ts
Comment (IgorNadj) [reply packages/data-broker/src/types/models.ts:91]: Cool, makes sense.

**Review by IgorNadj — CHANGES_REQUESTED** [review]
Nice one, thanks again for this, been wanting this for years!

File: packages/data-broker/src/types/models.ts
Comment (kael89) [reply packages/data-broker/src/types/models.ts:33]: Yeap, or even better convert `@tupaia/database` to TS too :wink:

File: packages/utils/src/object.js
Comment (kael89) [reply packages/utils/src/object.js:112]: Yeap you can get pretty decent type annotation with JsDocs only, see e.g. [phaser](https://github.com/photonstorm/phaser)

File: packages/data-broker/src/services/data-lake/DataLakeService.ts
Comment (kael89) [reply packages/data-broker/src/services/data-lake/DataLakeService.ts:79]: Oops yeah - great catch! :eyes:

File: packages/data-broker/src/services/dhis/types.ts
Comment (kael89) [reply packages/data-broker/src/services/dhis/types.ts:30]: Actually, I think this is a remnant from an earlier approach I used, where I was using

* `DataSourceI` as parent type for `DataElement,` DataGroup` etc
* `DataSource` as the type for `data_source` records

In the end I realised that this distinction was not required, so I decided to use a simpler approach for reasons similar to the ones you mention :slightly_smiling_face: 

I'll rename this to `DataSource`, nice catch!

File: packages/data-broker/src/services/dhis/types.ts
Comment (kael89) [reply packages/data-broker/src/services/dhis/types.ts:64]: No worries - it's great that the project is open source and I can still contribute in a way :slightly_smiling_face:

File: packages/data-broker/src/types/models.ts
Comment (rohan-bes) [reply packages/data-broker/src/types/models.ts:33]: Yeah that's fine about not re-using the types in `entity-server`, thanks for checking. They are slightly different. Ideally they'd both sit in `@tupaia/database` as you say 👍 

Just wanna note that if we do consider re-writing `@tupaia/database` in typescript, I'd love if we took the chance to seriously consider switching to a higher level ORM which has inbuilt typescript support. While I've always been impressed by the stability of our custom ORM; it's a lot of code, it quite hard to read through, and it could be more efficient/have more utility functions.

**Review by rohan-bes — COMMENTED** [review]
Thanks so much for this @kael89 !!! This is amazing!! 🎉 🎉 

I hope you're having an amazing time in Greece buddy! Apologies for taking so long to get to this, things have been pretty busy and then I got COVID, but I'm so glad to finally have read through it.

It's all looking really good honestly, and I appreciate not touching the functionality in this first pass. This is gonna make life a lot easier! (Next up `@tupaia/aggregator`?)

**Comment by IgorNadj** [comment]
We've had a great PR come through into dev to split up table data_source into data_element and data_group, which is causing all of these conflicts. I'm happy to resolve those if you like @kael89 ?

**Comment by kael89** [comment]
Thanks for the reviews @IgorNadj and @rohan-bes! I'll try to have a look at your comments and address all requested changes in the next few days. I will also try to put up a second PR that fixes any issues causing the build to fail.

Don't worry about the conflicts @IgorNadj , I'll have a look into them

File: packages/data-broker/src/types/models.ts
Comment (kael89) [reply packages/data-broker/src/types/models.ts:33]: Good point @rohan-bes ! I have tried https://www.prisma.io/ in a personal project and I was impressed by its TS support. I cannot do _everything_ (e.g. it doesn't support dynamic types for its CRUD operations - see [this GitHub issue](https://github.com/prisma/prisma/issues/5273)) but it's a great starting point.

File: packages/data-broker/src/__tests__/services/kobo/KoBoService.fixtures.ts
Comment (kael89) [reply packages/data-broker/src/__tests__/services/kobo/KoBoService.fixtures.ts:12]: That's right @rohan-bes

File: packages/data-broker/src/services/Service.ts
Comment (kael89) [reply packages/data-broker/src/services/Service.ts:41]: I belive that specific schemas for `options` should not be part of this `abstract` class contract. This is different from the return type: `Promise<AnalyticResults | EventResults | SyncGroupResults | undefined> | never` since we know that a service should return either events/analytics/sync groups or throw.

To phrase it in another way, the types for this class shouldn't have to be updated every time a new service is added.

That being said, I regret the decision for using `pull` instead of `pullAnalytics` and `pulEvents`: I think it is a wrong abstraction, in the same way that forcing data elements and data groups in the same table was a wrong abstraction, I think. (`Aggregator` has a better API in that regard).

File: packages/data-broker/src/services/dhis/DhisTranslator.ts
Comment (kael89) [reply packages/data-broker/src/services/dhis/DhisTranslator.ts:23]: Good point, perhaps `OutboundDataValue`?

File: packages/data-broker/src/services/dhis/pullers/DeprecatedEventsPuller.ts
Comment (kael89) [reply packages/data-broker/src/services/dhis/pullers/DeprecatedEventsPuller.ts:9]: I use `Partial` when all properties are de facto optional. This is a pattern we use often: the `options` object being entirely optional.

Happy to use the more explicit `?: ` syntax if you think it is clearer! Also tagging @rohan-bes for a second opinion

File: packages/data-broker/src/services/dhis/DhisService.ts
Comment (kael89) [reply packages/data-broker/src/services/dhis/DhisService.ts:250]: Oops good catch, will fix!

File: packages/data-broker/src/services/tupaia/TupaiaService.ts
Comment (kael89) [reply packages/data-broker/src/services/tupaia/TupaiaService.ts:1]: Aah that's a good trick, thanks for bringing it to my attention!!

File: packages/data-broker/src/services/weather/types.ts
Comment (kael89) [reply packages/data-broker/src/services/weather/types.ts:14]: The approach used here has some advantages:

1. Types do not have to be shared/redefined between higher and lower level modules. For example, `DataElement` in `dhis` supports a `dhisDataType` option, which is defined as an exported `type DataType = 'DataElement' | 'Indicator' | 'ProgramIndicator'`.
2. It is easier to see and a glance which options are supported by which service. This is also possible with a union type at the higher level types, it would just be more verbose.

On the flip side (also mentioned in your comment), listing all possible options is a more accurate representation of the actual data source records. There is a tradeoff here, I personally like the current pattern as it feels like type details are defined at the right level.

I'm definitely happy changing this, just wanted to communicate the aspects that influenced the current implementation before doing so. Please let me know what you think :-)

File: packages/data-broker/src/services/kobo/types.ts
Comment (IgorNadj) [reply packages/data-broker/src/services/kobo/types.ts:12]: These are Record<string, string>

File: packages/data-broker/src/services/dhis/pullers/DeprecatedEventsPuller.ts
Comment (IgorNadj) [reply packages/data-broker/src/services/dhis/pullers/DeprecatedEventsPuller.ts:9]: Happy either way :)

File: packages/data-broker/src/services/weather/types.ts
Comment (IgorNadj) [reply packages/data-broker/src/services/weather/types.ts:14]: Either way seems good to me. Pros and cons as you say. Happy with it as you have it now.

File: packages/data-broker/src/services/tupaia/TupaiaService.ts
Comment (kael89) [reply packages/data-broker/src/services/tupaia/TupaiaService.ts:1]: @IgorNadj I tried that and it works for individual commits, however those files still show separately in overall diff in GitHub....

It is a pity because it makes reviewing harder

File: packages/data-broker/src/types/results.ts
Comment (IgorNadj) [inline packages/data-broker/src/types/results.ts:9]: Minor: comment here would be nice for period property, should be in format YYYYMMDD e.g. "20200103"

Comment (kael89) [reply packages/data-broker/src/types/results.ts:9]: Good point, I'll add in the next PR :-)

**Comment by IgorNadj** [comment]
@kael89 merging dev in, resolving conflicts, please hold off on pushing to this branch

**Comment by kael89** [comment]
> @kael89 merging dev in, resolving conflicts, please hold off on pushing to this branch

I'm happy to rebase from `dev` and resolve conflicts if you want? I will just need some help with the CI/CD errors

**Comment by IgorNadj** [comment]
> > @kael89 merging dev in, resolving conflicts, please hold off on pushing to this branch
> 
> I'm happy to rebase from `dev` and resolve conflicts if you want? I will just need some help with the CI/CD errors

It's ok, I'm pretty familiar with latest changes in dev since I made most of them, happy to merge. Will have a look at CI/CD too, thanks though!

**Comment by kael89** [comment]
> > > @kael89 merging dev in, resolving conflicts, please hold off on pushing to this branch
> > 
> > 
> > I'm happy to rebase from `dev` and resolve conflicts if you want? I will just need some help with the CI/CD errors
> 
> It's ok, I'm pretty familiar with latest changes in dev since I made most of them, happy to merge. Will have a look at CI/CD too, thanks though!

No worries, let me know if you need any help! 🙌

**Comment by IgorNadj** [comment]
Replaced by https://github.com/beyondessential/tupaia/pull/4203

**Pull request closed without merging**
